# Autolib Dataset Cleaning

Work with the Autolib dataset for an electric car-sharing service company. Process stations data to understand electric car usage over time

To compute usage, we will need to understand that we have to join successive (in time) measures/counters for a given station, as the difference will tell whether a car was picked up, returned, or nothing happened.

## Importing our Libraries

In [0]:
# Importing the Pandas Library
#
import pandas as pd
import pandas_profiling

# Importing the Numpy Library
#
import numpy as np

## Reading the Dataset from our CSV file

In [0]:
url = ('http://bit.ly/autolib_dataset')
autolib = pd.read_csv(url)
autolib

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]

## Previewing our Dataset

In [0]:
autolib.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 25 columns]

## Accessing Information about our Dataset

In [0]:
autolib.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
Address                5000 non-null object
Cars                   5000 non-null int64
Bluecar counter        5000 non-null int64
Utilib counter         5000 non-null int64
Utilib 1.4 counter     5000 non-null int64
Charge Slots           5000 non-null int64
Charging Status        5000 non-null object
City                   5000 non-null object
Displayed comment      111 non-null object
ID                     5000 non-null object
Kind                   5000 non-null object
Geo point              5000 non-null object
Postal code            5000 non-null int64
Public name            5000 non-null object
Rental status          5000 non-null object
Scheduled at           47 non-null object
Slots                  5000 non-null int64
Station type           5000 non-null object
Status                 5000 non-null object
Subscription status    5000 non-null object
year                   

In [0]:
autolib.shape

(5000, 25)

## Cleaning our Dataset

In [0]:
#checking for unique values on the 	Displayed comment	column
autolib['Displayed comment'].unique()

array([nan, 'Station en parking (niv -1), accès 37 rue Leblanc',
       'Station en parking (niv 0), accès 26 villa Croix Nivert',
       'Station en parking (niv -2), accès 4 av. Foch',
       "Borne d'abonnement en sous-sol (niv -1), accès 4 rue Lobau",
       'Station en parking (niv -1), accès rue Wilson',
       'Station en parking (niv -1), accès Place Georges Pompidou',
       'Station en parking (niv -3), accès 32 rue dâ\x80\x99Alsace',
       'Station en parking (niv -2), accès 2 avenue Gabriel',
       'Station en parking (niv -1), accès av. de Versailles',
       'Station en parking (niv -1), accès rue Abel Gance',
       'Station en parking (niv -1), accès 4 av. des Ternes',
       'Station en parking (niv -1), Station B (sur la droite)',
       'Station en parking',
       'Station en parking (niv -2), accès 125 bvd du Montparnasse',
       'Station en parking (niv -1), Station A (sur la gauche)',
       'Station en parking (niv -1), accès 8 rue Louise Michel'

In [0]:
#dropping th Displayed comment column
autolib = autolib.drop(columns='Displayed comment')
autolib.head(2)

Address  Cars  Bluecar counter  ...  day  hour  minute
0        2 Avenue de Suffren     0                0  ...    8    11      43
1  145 Rue Raymond Losserand     6                6  ...    6     7      24

[2 rows x 24 columns]

In [0]:
autolib.describe()

Cars  Bluecar counter  ...         hour       minute
count  5000.000000      5000.000000  ...  5000.000000  5000.000000
mean      2.333600         2.333600  ...    11.509200    29.270000
std       2.035274         2.035274  ...     6.893549    17.231741
min       0.000000         0.000000  ...     0.000000     0.000000
25%       0.000000         0.000000  ...     6.000000    14.000000
50%       2.000000         2.000000  ...    11.000000    29.000000
75%       4.000000         4.000000  ...    18.000000    44.000000
max       7.000000         7.000000  ...    23.000000    59.000000

[8 rows x 12 columns]

In [0]:
#checking for unique values in the hour column
autolib['hour'].unique()

array([11,  7, 20,  4, 17, 18, 22, 15,  9, 12, 21,  8,  0, 19,  6,  3, 23,
        2,  1, 10, 16, 13, 14,  5])

In [0]:
#checking for unique values in month column
autolib['month'].unique()

array([4])

In [0]:
#checking for unique values in the year column
autolib['year'].unique()

array([2018])

In [0]:
#checking for unique values in the city column
autolib['City'].unique()

array(['Paris', 'Le Bourget', 'Nanterre', 'Puteaux', 'Fontenay-aux-Roses',
       'Le Chesnay', 'Suresnes', 'Versailles', 'Malakoff',
       'Issy-les-Moulineaux', 'Saint-Cloud', 'Sèvres', 'Bezons',
       'Nogent-sur-Marne', 'Créteil', 'Villejuif', 'Rueil-Malmaison',
       'Villeneuve-la-Garenne', 'Montreuil', 'Charenton-le-Pont',
       'Courbevoie', 'Asnières-sur-Seine', 'Garches', 'Drancy',
       'La Garenne-Colombes', 'Verrières-le-Buisson', 'Le Blanc-Mesnil',
       'Vanves', 'Enghien-les-Bains', 'Sceaux', 'Levallois-Perret',
       'Boulogne-Billancourt', 'Châtenay-Malabry', 'Pantin', 'Meudon',
       'Gennevilliers', 'Vincennes', 'Rungis', 'Colombes',
       'Fontenay-Sous-Bois', 'Rosny-sous-Bois', 'Bourg-la-Reine',
       'Cachan', 'Maisons-Alfort', 'Bièvres', 'Alfortville',
       'Champigny-sur-Marne', 'Aulnay-sous-bois', 'Saint-Denis',
       'Le Plessis-Robinson', 'Le Pré-Saint-Gervais', 'Clamart',
       'Bois-Colombes', 'Guyancourt', 'Bagnolet', 'Vaucresson',
  

In [0]:
#checking for unique values in the day column
autolib['day'].unique()

array([8, 6, 3, 4, 2, 1, 5, 9, 7])

In [0]:
autolib.isnull().sum()

Address                   0
Cars                      0
Bluecar counter           0
Utilib counter            0
Utilib 1.4 counter        0
Charge Slots              0
Charging Status           0
City                      0
ID                        0
Kind                      0
Geo point                 0
Postal code               0
Public name               0
Rental status             0
Scheduled at           4953
Slots                     0
Station type              0
Status                    0
Subscription status       0
year                      0
month                     0
day                       0
hour                      0
minute                    0
dtype: int64

In [0]:
#dropping month and year columns as they don't contain unique values
autolib.drop(['year', 'month'], axis=1, inplace=True)
autolib.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 22 columns]

In [0]:
#filling all null values with 0
autolib.fillna(0)

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 22 columns]

In [0]:
#checking for duplicated rows
autolib[autolib.duplicated()]

Empty DataFrame
Columns: [Address, Cars, Bluecar counter, Utilib counter, Utilib 1.4 counter, Charge Slots, Charging Status, City, ID, Kind, Geo point, Postal code, Public name, Rental status, Scheduled at, Slots, Station type, Status, Subscription status, day, hour, minute]
Index: []

In [0]:
#drop duplicated rows
autolib[autolib.duplicated(subset=None, keep= 'first')]

Empty DataFrame
Columns: [Address, Cars, Bluecar counter, Utilib counter, Utilib 1.4 counter, Charge Slots, Charging Status, City, ID, Kind, Geo point, Postal code, Public name, Rental status, Scheduled at, Slots, Station type, Status, Subscription status, day, hour, minute]
Index: []

In [0]:
autolib.to_csv('autolib1.csv')

In [0]:
# Load csv
import pandas as pd
df=pd.read_csv('autolib1.csv')

In [0]:
df.fillna(0)

Unnamed: 0                             Address  Cars  ...  day  hour  minute
0              0                 2 Avenue de Suffren     0  ...    8    11      43
1              1           145 Rue Raymond Losserand     6  ...    6     7      24
2              2    2 Avenue John Fitzgerald Kennedy     3  ...    3    20      14
3              3              51 Rue Eugène Oudiné     3  ...    4     4      37
4              4  6 avenue de la Porte de Champerret     3  ...    8    17      23
...          ...                                 ...   ...  ...  ...   ...     ...
4995        4995                 8 avenue Ménelotte     2  ...    6    11      26
4996        4996                   37 rue de Dantzig     4  ...    4    16      56
4997        4997                      142 rue du Bac     1  ...    1     7       1
4998        4998         2 avenue du Val de Fontenay     2  ...    4    17      27
4999        4999                 5 Rue Henry Brisson     4  ...    8    10       2

[5000 rows x 23 columns]

In [0]:
df.head()

Unnamed: 0                             Address  Cars  ...  day  hour  minute
0           0                 2 Avenue de Suffren     0  ...    8    11      43
1           1           145 Rue Raymond Losserand     6  ...    6     7      24
2           2    2 Avenue John Fitzgerald Kennedy     3  ...    3    20      14
3           3              51 Rue Eugène Oudiné     3  ...    4     4      37
4           4  6 avenue de la Porte de Champerret     3  ...    8    17      23

[5 rows x 23 columns]

In [0]:
df['Kind'].unique()

array(['STATION', 'PARKING', 'SPACE', 'CENTER'], dtype=object)

In [0]:
# Drop more unwanted columns
df = df.drop(['Unnamed: 0','Charge Slots','Charging Status','Rental status','Slots',
              'minute','Cars','Geo point','Scheduled at','ID'],axis=1)

In [0]:
df.head(2)

Address  Bluecar counter  ...  day  hour
0        2 Avenue de Suffren                0  ...    8    11
1  145 Rue Raymond Losserand                6  ...    6     7

[2 rows x 13 columns]

In [0]:
# Quick analysis of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
Address                5000 non-null object
Bluecar counter        5000 non-null int64
Utilib counter         5000 non-null int64
Utilib 1.4 counter     5000 non-null int64
City                   5000 non-null object
Kind                   5000 non-null object
Postal code            5000 non-null int64
Public name            5000 non-null object
Station type           5000 non-null object
Status                 5000 non-null object
Subscription status    5000 non-null object
day                    5000 non-null int64
hour                   5000 non-null int64
dtypes: int64(6), object(7)
memory usage: 507.9+ KB


In [0]:
#Check for completeness: missing values
df.isnull().sum()

Address                0
Bluecar counter        0
Utilib counter         0
Utilib 1.4 counter     0
City                   0
Kind                   0
Postal code            0
Public name            0
Station type           0
Status                 0
Subscription status    0
day                    0
hour                   0
dtype: int64

In [0]:
#What station is the most popular overall?
df[df.Kind == 'STATION'].groupby("Public name")["Public name"].count().sort_values(ascending=False).head()

Public name
Paris/Porte de Montrouge/8    13
Paris/Mathis/35               11
Paris/Philippe Auguste/126    11
Paris/Voltaire/182            11
Sèvres/Wolfenbüttel/1       11
Name: Public name, dtype: int64

Most Popular Station: Paris/Porte de Montrouge/8

In [0]:
# Rename the columns for uniformity
df = df.rename(columns={"Bluecar counter": 'bluecar_counter',
                        "Utilib counter": 'utilib_counter',
                        "Utilib 1.4 counter": 'utilib_14_counter',
                        "Rental status": 'rental_status'})

In [0]:
# Changing the case of column names to lower, stripping whitespaces, removing periods and replacing spaces with underscore
#
df.columns = df.columns.str.lower().str.strip().str.replace(' ', '_').str.replace('.', '')

# Confirmng that the dataframe columns have been renamed appriopriately
df.columns

Index(['address', 'bluecar_counter', 'utilib_counter', 'utilib_14_counter',
       'city', 'kind', 'postal_code', 'public_name', 'station_type', 'status',
       'subscription_status', 'day', 'hour'],
      dtype='object')

In [0]:
df.head()

address  bluecar_counter  ...  day  hour
0                 2 Avenue de Suffren                0  ...    8    11
1           145 Rue Raymond Losserand                6  ...    6     7
2    2 Avenue John Fitzgerald Kennedy                3  ...    3    20
3              51 Rue Eugène Oudiné                3  ...    4     4
4  6 avenue de la Porte de Champerret                3  ...    8    17

[5 rows x 13 columns]

In [0]:
# Adding a usage column for the three car counters i.e bluecar_counter, utilib_counter and utilib_14_counter
# creating column for bluecar usage
df['usage_blue'] = df.bluecar_counter.diff()

# creating column for utilib usage
df['usage_utilib'] = df.utilib_counter.diff()

# creating column for utilib 14 usage
df['usage_utilib_14'] = df.utilib_14_counter.diff()

positive number means that a car was returned \
negative number means that a car was picked\
zero means that nothing happened, i.e no car was picked nor returned

In [0]:
df.head()

address  ...  usage_utilib_14
0                 2 Avenue de Suffren  ...              NaN
1           145 Rue Raymond Losserand  ...              0.0
2    2 Avenue John Fitzgerald Kennedy  ...              2.0
3              51 Rue Eugène Oudiné  ...             -2.0
4  6 avenue de la Porte de Champerret  ...              0.0

[5 rows x 16 columns]

In [0]:
df[['bluecar_counter','usage_blue']]

bluecar_counter  usage_blue
0                   0         NaN
1                   6         6.0
2                   3        -3.0
3                   3         0.0
4                   3         0.0
...               ...         ...
4995                2         2.0
4996                4         2.0
4997                1        -3.0
4998                2         1.0
4999                4         2.0

[5000 rows x 2 columns]

## **Answering the Questions**

**Question 1** \

Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018

In [0]:
# Unique bluecar usage
df.usage_blue.unique()

In [0]:
# Most popular hour
# usage is negative when picking up cars
df[(df.city == 'Paris') & 
   (df.usage_blue < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)

Most popular hour is 2100 Hrs

In [0]:
# Alternative
# Subset city to find Paris only
paris_df = df[df.city == "Paris"]
paris_df[['bluecar_counter','usage_blue','hour']]

bluecar_counter  usage_blue  hour
0                   0         NaN    11
1                   6         6.0     7
3                   3         0.0     4
4                   3         0.0    17
5                   0        -3.0     7
...               ...         ...   ...
4992                0        -5.0    23
4994                0        -6.0    20
4996                4         2.0    16
4997                1        -3.0     7
4999                4         2.0    10

[2816 rows x 3 columns]

In [0]:
# Alternative
# Usage is -ve when picked
paris_df[(paris_df.usage_blue < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)

hour
21    80
Name: hour, dtype: int64

**Question 2** \
What is the most popular hour for returning cars?


In [0]:
# Most popular hour for returning cars
# usage is positive when returning
df[(df.city == 'Paris') & 
   (df.usage_blue > 0)].groupby('hour')['usage_blue'].count().sort_values(ascending = False).head(1)

Most popular hour is 0300 Hrs

**Question 3** \
What station is the most popular \
a.) Overall?

In [0]:
# Identifying the station that is the most popular overall

df[(df.kind == 'STATION') & 
   (df.status == 'ok')].groupby('public_name')[['kind']].count().sort_values(by = 'kind', ascending  = False).head(1)

kind
public_name                     
Paris/Porte de Montrouge/8    13

Overall most popular station is **Paris/Porte de Montrouge/8**

b.) At the most popular picking hour?

In [0]:
# Identifying the station that is the most popular at the most popular picking hour
#
df[(df.kind == 'STATION') & (df.hour == 21)  & 
   (df.status == 'ok')].groupby('public_name').count()[['kind']].sort_values(by = 'kind', ascending = False).head(1)

kind
public_name            
Paris/Tronchet/19     4

Most popular station at the most popular hour is **Paris/Tronchet/19**


**Question 4** \
What postal code is the most popular for picking up Bluecars? Does the most popular station belong to that postal code? \
a.) Overall?

In [0]:
# Identifying the postal code which is the most popular for picking up Bluecars - overall
# Usage is negative when picking up cars
#
df[df.usage_blue < 0].groupby('postal_code').count()[['usage_blue']].sort_values(by = 'usage_blue', ascending = False).head(1)

usage_blue
postal_code            
75015               116

In [0]:
# Checking whether the most popular station belongs to the most popular postal code for picking up Bluecars 
# Overall
#
df[df.public_name == 'Paris/Porte de Montrouge/8'][['public_name', 'postal_code']].head(1)

public_name  postal_code
122  Paris/Porte de Montrouge/8        75014

b.) At the most popular picking hour?

In [0]:
#Identifying the postal code which is the most popular for picking up Bluecars - At the most popular picking hour
# Usage is negative when picking up cars
#
df[(df.usage_blue < 0) & 
   (df.hour == 21)].groupby('postal_code').count()[['usage_blue']].sort_values(by = 'usage_blue', ascending = False).head(1)

usage_blue
postal_code            
75008                 9

In [0]:
# Checking whether the most popular station belongs to the most popular postal code for picking up Bluecars 
# At the most popular picking hour?
#
df[(df.public_name == 'Paris/Tronchet/19') & (df.hour == 21)][['public_name', 'postal_code']].head(1)

public_name  postal_code
2172  Paris/Tronchet/19        75008

**Question 5** \
Do the results change if you consider Utilib and Utilib 1.4 instead of Bluecars? 

**Utilib Analysis**

In [0]:
# Checking whether the utilib_counter had any picking or returning of cars
#
df.usage_utilib.unique()

In [0]:
# Identifying the most popular hour of the day for picking up a shared electric car (Utilib)
# in the city of Paris over the month of April 2018
# Usage is negative when picking up cars
#
df[(df.city == 'Paris') & (df.usage_utilib < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)



> The most popular hour of the day for picking up a shared electric car (Utilib) 
in the city of Paris over the month of April 2018 was 1900Hrs



What is the most popular hour for returning cars ?

In [0]:
# Identifying the the most popular hour for returning cars (Utilib)
# Usage is positive when returning cars
#
df[df.usage_utilib > 0].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)



> The most popular hour for returning cars from Utilib is 0500Hrs



What station is the most popular overall?


In [0]:
df[(df.kind == 'STATION') & 
            (df.status == 'ok')].groupby('public_name')[['kind']].count().sort_values(by = 'kind', ascending  = False).head(1)

What is the most popular station at the most popular hour?

In [0]:
df[(df.kind == 'STATION') & (df.hour == 19)  & 
            (df.status == 'ok')].groupby('public_name').count()[['kind']].sort_values(by = 'kind', ascending = False).head(1)

What is the most popular hour for returning cars? \
Overall?

In [0]:
# Identifying the postal code which is the most popular for picking up Utilib - overall
# Usage is negative when picking up cars
#
df[df.usage_utilib < 0].groupby('postal_code').count()[['usage_utilib']].sort_values(by = 'usage_utilib', ascending = False).head(1)



> Postal Code for most popular for picking up cars is 75015



What is the most popular hour for returning cars? \
At the most popular picking hour?

In [0]:
# Identifying the postal code which is the most popular for picking up Utilib cars - At the most popular picking hour
# Usage is negative when picking up cars
#
df[(df.usage_utilib < 0) & 
   (df.hour == 19)].groupby('postal_code').count()[['usage_utilib']].sort_values(by = 'usage_utilib', ascending = False).head(1)

> Postal Code for most popular for picking up at the most popular hour is 75014

Does the most popular station belong to that postal code overall?

In [0]:
df[df.public_name == 'Paris/Porte de Montrouge/8'][['public_name', 'postal_code']].head(1)

In [0]:
df[df.public_name == "Alfortville/Port à l'Anglais/39"][["public_name", "postal_code"]].head(1)


In [0]:
# Checking whether the utilib_14 had any picking or returning of cars
#
df.usage_utilib_14.unique()

In [0]:
# Identifying the most popular hour of the day for picking up a shared electric car (Utilib 1.4)
# in the city of Paris over the month of April 2018
# Usage is negative when picking up cars
#
df[(df.city == 'Paris') & (df.usage_utilib_14 < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)

The most popular hour of the day for picking up a shared electric car (Utilib 
1.4) in the city of Paris over the month of April 2018 was 0300Hrs

The most popular hour for returning cars from Utilib is 0500Hrs

In [0]:
# Identifying the the most popular hour for returning cars (Utilib 1.4)
# Usage is positive when returning cars
#
df[df.usage_utilib_14 > 0].groupby('hour')['hour'].count().sort_values(ascending = False).head(1)

The most popular hour for returning cars from Utilib 14 is 0300Hrs

What postal code is the most popular for picking up an electric car? \
Overall?

In [0]:
# Identifying the postal code which is the most popular for picking up Utilib 1.4- overall
# Usage is negative when picking up cars
#
df[df.usage_utilib_14 < 0].groupby('postal_code').count()[['usage_utilib_14']].sort_values(by = 'usage_utilib_14', ascending = False).head(1)

What postal code is the most popular for picking up cars? \
At the most popular picking hour?

In [0]:
# Identifying the postal code which is the most popular for picking up Utilib 14 cars - At the most popular picking hour
# Usage is negative when picking up cars
#
df[(df.usage_utilib_14 < 0) & 
   (df.hour == 3)].groupby('postal_code').count()[['usage_utilib_14']].sort_values(by = 'usage_utilib_14', ascending = False).head(1)